Adding material mixing to openmc.Material class

eepeterson opened the pull request:
This is just a draft right now, but it's a simple PR to add some functionality to the Material class based on the stale PR #773. The main capabilities are the ability to add a material to another material and to mix an iterable of materials by volume fractions.

paulromano commented on openmc/material.py:
```diff
@@ -560,6 +560,73 @@ def add_element(self, element, percent, percent_type='ao', enrichment=None):
             self.add_nuclide(*nuclide)
 
+    def add_material(self, material, percent, percent_type='ao'):
+        """Mix a material to the material
```
Maybe "Mix a material into this material"? It's a little unclear as is. Also, I would recommend having a paragraph (after the one-liner but before Parameters) that describes this in a little more detail. In particular, how is this mixing performed? Are the nuclides from the other material added in (increasing the density of the current material), or does everything scale so that the density is preserved? Right now, a user won't be able to figure out without reading the source code.

paulromano commented on openmc/material.py:
```diff
@@ -560,6 +560,73 @@ def add_element(self, element, percent, percent_type='ao', enrichment=None):
+                  'different temperature "{1}"'.format(self.id, temperature)
+            raise ValueError(msg)
+
+        elif percent_type not in ['ao', 'wo']:
```
```suggestion
        elif percent_type not in ('ao', 'wo'):
```

paulromano commented on openmc/material.py:
```diff
@@ -560,6 +560,73 @@ def add_element(self, element, percent, percent_type='ao', enrichment=None):
+        """
+
+        if self._macroscopic is not None:
+            msg = 'Unable to add a Material to Material ID="{0}" as a ' \
```
Numbered format is not necessary here:
```suggestion
            msg = 'Unable to add a Material to Material ID="{}" as a ' \
```
Same thing below

paulromano commented on openmc/material.py:
```diff
@@ -560,6 +560,73 @@ def add_element(self, element, percent, percent_type='ao', enrichment=None):
+        # Add the nuclide to the material
+        for nuc, dens in new_nucs.values():
+            density = dens / density_sum
+            self.add_nuclide(nuc, density*percent, percent_type)
```
If the nuclide is already present in the material, this will result in the material having that nuclide duplicated. It would be better to modify the density of the existing nuclide.

paulromano commented on openmc/material.py:
```diff
@@ -560,6 +560,73 @@ def add_element(self, element, percent, percent_type='ao', enrichment=None):
+
+        # Add the s(a,b) tables
+        for sab in material.sab:
+            self.add_s_alpha_beta(sab)
```
Again, we'll need a check for an existing S(a,b) table, and the appropriate fraction should be applied.

eepeterson replied:
@paulromano Could you take another look at this? I'm trying to figure out how to ensure the correct thermal scattering cross sections are added when mixing materials. I don't see a method for finding out what nuclides are in a particular S(a,b) table and if multiple nuclides are in a particular table like 'c_BeO' or 'c_C6H6' how are the fractions applied? It seems like things get messy quickly when multiple S(a,b) tables are being mixed with materials that might have the same nuclides but not with S(a,b) tables. Do you have any advice on how to proceed with this?

paulromano replied:
Ugh, yeah I see what you mean. It does start to get messy to cover all the possible scenarios. So as far as figuring out what nuclides are relevant for a particular table, check out [this logic](https://github.com/openmc-dev/openmc/blob/de99faefc93fe46a48fb4c5abcb4b88fe0eec174/openmc/data/thermal.py#L718-L728) which almost solves that. I say _almost_ because you would need to know what the appropriate atomic number is and we don't really have that information at hand.

A reasonable approach would be to look at the first applicable nuclide in the materials being mixed and use the relative densities of that nuclide to determine what fraction of the S(a,b) table to apply. Note that the fractions do not need to sum to unity; say you were mixing two materials that were exactly the same, but one of them used an S(a,b) table. In that case, you would just have `fraction=0.5` on the S(a,b) table for the resulting mixed material.

Another reasonable approach would be to just punt on this and throw an exception saying that we don't support mixing materials with S(a,b) tables :laughing: Not sure if that affects your intended use case.

eepeterson replied:
Haha, ok since it doesn't affect my use case yet, I'm going to punt and raise a `NotImplementedError` for now.

paulromano commented on openmc/material.py:
```diff
@@ -914,6 +981,52 @@ def to_xml_element(self, cross_sections=None):
         return element
 
+    @classmethod
+    def volume_mix(materials,fracs,name=None):
```
```suggestion
    def volume_mix(materials, fracs, name=None):
```

paulromano commented on openmc/material.py:
```diff
@@ -914,6 +981,52 @@ def to_xml_element(self, cross_sections=None):
+
+        """
+        fracs = np.asarray(fracs)
+        if np.sum(fracs) != 1:
```
You might want to just check if it is sufficiently close to 1. Floating point equality is difficult to ensure.

paulromano commented on openmc/material.py:
```diff
@@ -914,6 +981,52 @@ def to_xml_element(self, cross_sections=None):
+        fracs = np.asarray(fracs)
+        if np.sum(fracs) != 1:
+            # Issue warning that fractions don't add to 1 set remainder to void
+            print("Warning: sum of volume fractions do not add to 1, \
```
Use `warning.warn` from the standard library

paulromano commented on openmc/material.py:
```diff
@@ -914,6 +981,52 @@ def to_xml_element(self, cross_sections=None):
+            print("Warning: sum of volume fractions do not add to 1, \
+                  void fraction set to {}".format(1-np.sum(fracs)))
+        if name is None:
+            name = "-".join(["{0}({1})".format(m.name,f) for m,f in zip(materials,fracs)])
```
Follow [PEP8](https://www.python.org/dev/peps/pep-0008/) for Python code
```suggestion
            name = "-".join(["{}({})".format(m.name, f) for m, f in zip(materials, fracs)])
```

paulromano commented on openmc/material.py:
```diff
@@ -914,6 +981,52 @@ def to_xml_element(self, cross_sections=None):
+        # Compute atom fractions of nuclides and add them to the new material  
+        new_density = 0.0
+        for nuc, atom_dens in nuclides_per_cc.items():
+            new_mat.add_nuclide(nuc,atom_dens/tot_nuclides_per_cc,"ao")
```
```suggestion
            new_mat.add_nuclide(nuc, atom_dens/tot_nuclides_per_cc, "ao")
```

paulromano commented on openmc/material.py:
```diff
@@ -914,6 +981,52 @@ def to_xml_element(self, cross_sections=None):
+            new_mat.add_nuclide(nuc,atom_dens/tot_nuclides_per_cc,"ao")
+            new_density += mass_per_cc[nuc]
+
+        new_mat.set_density("g/cm3",new_density)
```
```suggestion
        new_mat.set_density("g/cm3", new_density)
```

paulromano commented on openmc/material.py:
```diff
@@ -914,6 +981,52 @@ def to_xml_element(self, cross_sections=None):
+
+        # Compute sum of nuclide densities in combined material
+        dens_array = np.asarray([dens for dens in nuclides_per_cc.values()])
+        tot_nuclides_per_cc = np.sum(dens_array)
```
I would combine these two lines:
```Python
tot_nuclides_per_cc = np.sum([dens for dens in nuclides_per_cc.values()])
```

paulromano commented on openmc/material.py:
```diff
@@ -914,6 +981,52 @@ def to_xml_element(self, cross_sections=None):
+        mass_per_cc = defaultdict(float)
+        for mat, frac in zip(materials, fracs):
+            for nuc, atoms_per_bcm in mat.get_nuclide_atom_densities().values():
+                nuc_per_cc = frac*1.E24*atoms_per_bcm
```
```suggestion
                nuc_per_cc = frac*1.e24*atoms_per_bcm
```

eepeterson commented:
Thanks for the comments @paulromano. After thinking about it a little more, it doesn't make as much sense to me to have an add_material method since the ambiguity in the meaning might be too much. I've deleted that method and implemented just a class method for mixing multiple materials by volume, weight, or molar fractions. This is still a draft and I think I need to handle the addition of voids a little more carefully as well as write some tests.

paulromano commented on openmc/material.py:
```diff
@@ -914,6 +915,97 @@ def to_xml_element(self, cross_sections=None):
+        sab_names = set(sab[0] for mat in materials for sab in mat._sab)
+        if sab_names:
+            msg = 'Currently we do not support mixing materials containing \
+                S(a,b) tables'
```
This will result in a bunch of extra whitespace in the middle of the string. Better to just use implicit concatenation with something like:
```Python
msg = ('some very long string '
       'that continues over multiple lines')
```
When Python sees two string literals adjacent to one another, it joins them together

paulromano commented on openmc/material.py:
```diff
@@ -914,6 +915,97 @@ def to_xml_element(self, cross_sections=None):
+        # Warn that fractions don't add to 1, set remainder to void, or raise
+        # an error if percent_type isn't 'vo'
+        if not np.isclose(void_frac, 0.):
+            if percent_type in ['ao', 'wo']:
```
```suggestion
            if percent_type in ('ao', 'wo'):
```

paulromano commented on openmc/material.py:
```diff
@@ -914,6 +915,97 @@ def to_xml_element(self, cross_sections=None):
+        if not np.isclose(void_frac, 0.):
+            if percent_type in ['ao', 'wo']:
+                msg = 'A non-zero void fraction is not acceptable for \
+                    percent_type: {}'.format(percent_type)
```
Same issue here regarding whitespace in the middle of the string (see comment at line 979)

paulromano commented on openmc/material.py:
```diff
@@ -914,6 +915,97 @@ def to_xml_element(self, cross_sections=None):
+            Materials to combine
+        fracs : Iterable of float
+            Fractions of each material to be combined
+        percent_type : str
```
```suggestion
        percent_type : {'ao', 'wo', 'vo'}
```

paulromano commented on openmc/material.py:
```diff
@@ -914,6 +915,97 @@ def to_xml_element(self, cross_sections=None):
+
+        Returns
+        -------
+        new_mat : openmc.Material instance for the new mixture
```
I would just have
```rst
Returns
-------
openmc.Material
    Mixture of the materials
```
Since the name of a return value is optional

eepeterson commented:
@paulromano thanks for the edits, it should be good to go now once the checks complete
